notification-forwarder/notification-forwarder.py

```diff
@@ -0,0 +1,464 @@
+    notifications to other VMs. It may also decide to let the locally running
+    desktop notification server [1] handle incoming notifications depending on
+    e.g. the capabilities required by the requesting application.
+
+    To keep it simple and more secure, there's currently no channel back from the
+    receiving VM to this proxy (otherwise the receiving VM might use that to compromise
+    us).
+    In particular if the user closes or clicks a button on a forwarded notification, the
+    originating application won't be informed. Therefore the current implementation does
+    not forward notifications with buttons, but lets the locally running desktop
+    notification server handle them.
+
+    `dbus-monitor "interface=org.freedesktop.Notifications"` is useful for debugging.
+
+    The proxy assumes that the target VM to forward notifications to is named at the
+    QubesDB path `/desktop-notification-target`. Otherwise it will exit and let the
+    local desktop notification server handle all notifications.
+
+    References:
+        [1] https://developer.gnome.org/notification-spec/
+        [2] https://dbus.freedesktop.org/doc/dbus-python/
+        [3] https://lazka.github.io/pgi-docs/
+    """
+
+    def __init__(self, notification_vm, private_dbus_address, verbose=False, local_mode=False, force_mode=False):
+        self._id = 0 #: current server notification ID
+        self.notification_vm = notification_vm
+        self.local_mode = local_mode
+        self.force_mode = force_mode
+        self.log = get_logger("qubes-notification-forwarder", verbose)
+
+        self.server2local_id = dict() #: server notification ID --> local client ID
+        self.local2server_id = dict() #: local client ID --> server notification ID
+
+        dbus.mainloop.glib.DBusGMainLoop(set_as_default=True)
+
+        self.private_bus = dbus.bus.BusConnection(private_dbus_address)
+
+        #signal handling
+        bn = "org.freedesktop.Notifications"
+        path = "/org/freedesktop/Notifications"
+        self.private_bus.add_signal_receiver(self.on_local_closed, signal_name="NotificationClosed", bus_name=bn, path=path)
+        self.private_bus.add_signal_receiver(self.on_local_action, signal_name="ActionInvoked", bus_name=bn, path=path)
+
+        self.capabilities = self.iface.GetCapabilities()
+
+        self.qrexec_queue = Queue(100)
+        self.qrexec_worker = QrexecWorker(self.qrexec_queue, notification_vm, verbose=verbose)
+        self.qrexec_worker.start()
+
+        bus_name = dbus.service.BusName(bn, bus=dbus.SessionBus())
+        super().__init__(bus_name, path)
+
+    @property
+    def iface(self):
+        """ Interface for local notification handling. """
+        # Storing the proxy or interface doesn't seem to be wise: After a while (30s-10min) one runs into a
+        # DBus.Error.ServiceUnknown (i.e. some timeout) - at least on debian 10. Maybe a python-dbus bug?
```

Comment:
I recommend reporting this, or using a lower-level API.

Reply:
Not sure where it could be reported and whether I want this to waste my time. Anyway the workaround worked well during my tests and shouldn't bee too expensive.